### ISSUE TITLE
multiprocessing on dask.dataframe becomes inefficent if a column is added at runtime

### ISSUE THREAD

**Issue by abast** [dask/dask#1590]
Let dummy be a large dataframe (about 100GB of data). I noticed, that adding a column to that dataframe slows down all further computations.
```
dummy['cell_type'] = dummy.apply(lambda x: x.synapse_type.split('_')[0], axis = 1)
result = dummy.map_partition(my_fun).compute(get=dask.multiprocessing.get)
```

my_fun does not use the added column in any way. The above code results in 30% CPU usage and a long running time. With line 1 commented out, I get 100% CPU usage. 

Background information:
- The dataframe is generated with the .from_delayed method.
- The data is stored on a locally
- The compute server has 40 cores and 512GB memory.
- looking at `top` (linux command), there is usually one python process using slightly more than 100% CPU, while there are many python processes using about 20-30%. If line 1 is commented, all processes use ~95% CPU.

Does the additional column introduce heavy interprocess communication?  What bottlenecks are likely? Might this be connected to https://github.com/dask/dask/issues/1588?

**Comment by jcrist** [comment]
- Creating a new column shouldn't result in increased communication - the graph is still embarrassingly parallel along the partitions (you can see this using the `visualize` method).

- For many cases, it's actually better to use a local instance of the distributed scheduler instead of the multiprocessing scheduler - if for nothing else than that it's easier to debug. This is easy to setup:

```python
from distributed import client
c = Client()
# The distributed scheduler will now be set as the global default.
# If you don't want this, pass in set_as_default=False to Client,
# and specify the scheduler manually as `get=c.get`
# It should have as many worker processes as cores on your machine.
# If you'd like fewer processes, but a few threads in each process, you can
# specify that using `LocalCluster` (done automatically above):
from distributed.deploy import LocalCluster
cluster = LocalCluster(n_workers=10, threads_per_worker=4)
c = Client(cluster)
```

- If using the distributed scheduler, profiling using the [web interface](http://distributed.readthedocs.io/en/latest/web.html) should be useful in determining the cause of slowdowns. For the threaded/multiprocessing scheduler, the [profiling](http://dask.pydata.org/en/latest/diagnostics.html#profiling) capabilities can be used (specifically local at the `Profiler` and `ResourceProfiler`).

- Just to clarify, even though `my_fun` doesn't use the added column in any way, dask has no way to know if this is true. `map_partitions` applies a function to each partition - it doesn't apply it to only the columns needed. The same goes for `apply` - even though you're only using one column, the whole frame will be needed here.

- The psutil thing is unrelated. That's for determining the default chunk size when using `read_csv`.

**Comment by abast** [comment]
Thanks for your reply. Is it somehow possible to access the documentation of dask 0.10.2 and distributed 1.12.2? `Client ` seems to be a new feature, which does not exist in versions of distributed, that are compatible with dask 0.10.2, so I'm not sure how to set up the distributed scheduler.

With dask from the new master branch, I get some strange errors (see below), and many many of my unittests are broken, so to focus on actual analysis, I tried to postpone the upgrade.


[xcb] Unknown sequence number while processing queue
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
python: ../../src/xcb_io.c:274: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.

**Comment by jcrist** [comment]
> Is it somehow possible to access the documentation of dask 0.10.2 and distributed 1.12.2?

No, unfortunately you'd have to checkout a tag on the git repo and look at them yourself, we only keep the latest documentation on readthedocs.

> `Client` seems to be a new feature, which does not exist in versions of distributed, that are compatible with dask 0.10.2, so I'm not sure how to set up the distributed scheduler.

`Client` is the new name for `Executor` (renamed in the latest release). Just use `Executor` instead in the code above and everything should work fine.

> With dask from the new master branch, I get some strange errors (see below), and many many of my unittests are broken, so to focus on actual analysis, I tried to postpone the upgrade.

There were some breaking changes in the latest release, but the errors you posted look completely unrelated to dask. I'd expect those are related to some other library you may be using.

**Comment by abast** [comment]
Thanks again. Yes, I agree. However this error is 100% correlated to upgrading and downgrading dask, which I tried several times. Nevertheless, I hope to get some insights what the bottleneck might be using the distributed scheduler.

**Comment by abast** [comment]
About the probably unrelated error: I was able to use matplotlib in processes launched by dask in 0.10.2. In the current dask release, the matplotlib backend specified before is not used in the spawned processes anymore, falling back to the default backend, which causes the error. 

While the idea of using matplotlib within dask might seem strange, it gave me a huge speedup to plot data in parallel, and overlay them in the end.

**Comment by abast** [comment]
About the strange error: Apparently, matplotlib does not work as in dask 0.10.2 in processes launched by the multiprocessing scheduler. Matplotlib uses the standard backend instead of the one specified before, which causes the error. While using matplotlib in this way might seem strange, it provided a huge speedup by allowing to plot parts of the dataset separately and overlaying them in the end.

edit: using `Executor()` fails with an `OSError: [Errno 2] No such file or directory` in some bokeh script, even after updating bokeh. Is there some special version required?

**Comment by abast** [comment]
In the meantime some profiling results:
## With line 1 commented out, no added column:
![profiling_line1_commented_out](https://cloud.githubusercontent.com/assets/20513804/18975966/70894026-86ae-11e6-9d9e-b2003c3539c6.png)

 - blue: my_fun(read_csv(*))

Comment: This is as I expected. The machine has 40 cores reflected by 39 worker IDs. The actual computation starts immediately and I see ~95% CPU usage in the task manager.

## line 1 not commented out, column added at runtime, calling _, read_csv and head.

![profiling_with_line_1](https://cloud.githubusercontent.com/assets/20513804/18975967/7089a048-86ae-11e6-85b9-aa66018a5969.png)

 - blue: applying the anonymous function
 - green: applying my_fun
 - yellow: read_csv

Comment: The actual computatation is delayed for ~20s. There are 3 more processes than cores. Those cores (ID 40, 41, 42) are active in the first milliseconds.

**Comment by mrocklin** [comment]
Readthedocs is now hosting some older versions of our documentation: http://distributed.readthedocs.io/en/1.12.2/

**Comment by abast** [comment]
Moving all the csv files to tempfs didn't speed anything up, so apparently reading in the data is not the bottleneck. I am aware, that a minimal example is necessary. Should we close the issue until I can provide one?

**Comment by jcrist** [comment]
Apologies for the lack of response.

> Using Executor() fails with an OSError: [Errno 2] No such file or directory in some bokeh file, even after updating bokeh. Is there some special version required?

Shouldn't be. What's the full error message? You may also have a better experience debugging spinning up a cluster separately instead:

```bash
$ # In one terminal. This will print out a scheduler ip & port, use that below:
$ dask-scheduler
$ # In another terminal
$ dask-worker --nprocs=40 --nthreads=1 SCHEDULER_ADDRESS_FROM_ABOVE
```

> In the meantime some profiling results:

From looking at your profiling plot, it appears to be an increase in serialization cost causing the issue. The local profiler lumps the cost of calling a function and serializing the data for it together, so I can't be sure. The distributed profiler will display the serialization cost in red, so that would be more beneficial for verifying.

I'd expect the multiprocessing scheduler to not move data between processes, as it's a simple map operation going on here. However, some heuristic may be failing. If you can, please try your problem with the distributed scheduler. I expect it to be more efficient across multiple processes, and also should make debugging and profiling easier.

> Should we close the issue until I can provide one?

Only if you want to. I'm fine with keeping issues like this open as long as some activity is made to resolve them. I hope we can help you figure things out.

**Comment by jcrist** [comment]
> From looking at your profiling plot, it appears to be an increase in serialization cost causing the issue.

To verify this, try doing the following locally (assuming you're using IPython):

```python
# let `df` be a dataframe chunk of appropriate size.
# You can use dask to compute one locally if you're not
# sure what this is:
In [1]: df = dummy.get_partition(0).compute()

In [2]: df2 = df.assign(cell_type=df.apply(lambda x: x.synapse_type.split('_')[0], axis = 1))

In [3]: import cloudpickle

In [4]: with_col = %timeit -o cloudpickle.dumps(df2)

In [5]: without_col = %timeit -o cloudpickle.dumps(df)

In [6]: with_col.best - without_col.best
```

If the serialization cost is the problem, I'd expect the output of cell 6 to be relatively large.

**Comment by abast** [comment]
Thanks for your reply. After updating to dask 11.1, the delay before the actual computation begins is gone.
Given the following code, it appears, that the element-wise apply function is the bottleneck, so I close the issue.

    abbr = lambda: np.random.randint(1000, size = 2000000)
    pdf = pd.DataFrame(dict(A=abbr(), B=abbr(), c=abbr(), D=abbr()))
    #untouched ddf
    ddf = dd.from_pandas(pdf, npartitions = 400)
    #assignment to column
    ddf_assigned = dd.from_pandas(pdf, npartitions = 400)
    ddf_assigned['Z'] = ddf_assigned.apply(lambda x: x.A, axis = 1)
    #pandas --> dask --> assign --> pandas --> dask
    ddf_assigned_from_pandas = dd.from_pandas(ddf_assigned.compute(get = dask.multiprocessing.get), npartitions = 400)

    def fun(df):
        time.sleep(0.01)
        return pd.Series(dict(A=1))

    %time silent = ddf.map_partitions(fun, meta = int).compute(get = dask.multiprocessing.get)
    %time silent = ddf.apply(lambda x: x.A, axis = 1).compute(get = dask.multiprocessing.get)
    %time silent = ddf_assigned.map_partitions(fun, meta = int).compute(get = dask.multiprocessing.get)
    %time silent = ddf_assigned_from_pandas.map_partitions(fun, meta = int).compute(get = dask.multiprocessing.get)

This gives the following runtimes:
CPU times: user 560 ms, sys: 396 ms, total: 956 ms
Wall time: 1.43 s
CPU times: user 628 ms, sys: 236 ms, total: 864 ms
Wall time: 5.82 s
CPU times: user 1.63 s, sys: 420 ms, total: 2.05 s
Wall time: 7.73 s
CPU times: user 720 ms, sys: 316 ms, total: 1.04 s
Wall time: 1.45 s

 I am not
